scripts/updateMessages.js

```diff
@@ -1,51 +1,62 @@
-const fetchSpeechToText = async (url) => {
-  // getting translated message from storage.local
-  let savedUrls = await browser.storage.local.get();
-
-  if (savedUrls) {
-    for (let savedUrl in savedUrls) {
-      if (savedUrl === url) {
-        return savedUrls[savedUrl];
-      }
-    }
+/**
+ * A higher order function that provides caching in the browser's local storage.
+ *
+ * Only supports functions with a single argument that can be stored as an object's key
+ * (string, number, symbol, boolean).
+ */
+const withBrowserLocalCache = (fn) => (arg) => {
+  /** @type {Record<any, any>} */
+  const cache = (await browser.storage.local.get()) || {};
+  if (cache[arg]) {
+    return cache[arg];
   }
 
-  // fetching new data
+  const result = await fn(arg);
+  cache[arg] = result;
+  await browser.storage.local.set(cache);
+
+  return result;
+};
+
+const fetchSpeechToText = async (url) => {
   // TODO: fetch sererUrl from heroku server
   const serverUrl = "http://61e2-78-11-176-160.ngrok.io";
 
-  const headers = {
+  const requestOptions = {
     method: "POST",
     headers: { "Content-Type": "application/json" },
-    body: JSON.stringify({ url: url }),
+    body: JSON.stringify({ url }),
   };
 
-  const response = await fetch(serverUrl + "/speech-to-text", headers);
+  const response = await fetch(serverUrl + "/speech-to-text", requestOptions);
 
-  const message = await response.text();
-
-  if (savedUrls) {
-    savedUrls[url] = message;
-    browser.storage.local.set(savedUrls);
```

Comment:
While saving the data to the cache generally works, it has a data race that could potentially result in some messages not being cached :smile: 

Imagine the following:
1. A new audio message appears. Let's call it A. `savedUrls` is read from the cache and a request is sent.
2. Another audio message appears. Let's call it B. `savedUrls` is again read from the cache (it's the same as `savedUrls` for message A) and a request is sent.
3. The response for A arrives. The cache is updated (it's the initial cache + the result for A).
4. The response for B arrives. The cache is updated - but here, it does not check for the latest cache. It still uses the `savedUrls` from before making the request. It does not know that a new result for A has been added to the cache. Thus, it overwrites the cache - the cached URLs are now the initial ones + the result for B. The result for A is overwritten

I thought I fixed it, but looks like I did not not. A proper fix would be to use some kind of synchronization, but that's form over substance :smile: The "good enough" fix would be to reread the cache just before updating it.

Reply:
That is a great remark. I will add this bug to my TODO list and fix it as soon as possible. Not cached messages will cause more requests to the server which is terrible for me.